File: backend/app/schemas/batch.py
from typing import List, Optional, Any, Dict
from datetime import datetime
from pydantic import BaseModel, ConfigDict, Field
import uuid
from typing import Literal


class BatchJobCreate(BaseModel):
    operation_type: Literal["analyze", "score", "tag", "archive"]
    job_ids: List[uuid.UUID] = Field(..., min_length=1, max_length=500)
    params: Dict[str, Any] = Field(default_factory=dict)


class BatchJobCreateResponse(BaseModel):
    id: uuid.UUID
    operation_type: str
    status: str
    total_items: int
    processed_items: int
    succeeded_items: int
    failed_items: int
    payload: Dict[str, Any]
    result_summary: Dict[str, Any]
    cancel_requested: bool
    started_at: Optional[datetime]
    completed_at: Optional[datetime]
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class BatchJobUpdate(BaseModel):
    cancel_requested: Optional[bool] = None


class BatchJobResponse(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    operation_type: str
    status: str
    total_items: int
    processed_items: int
    succeeded_items: int
    failed_items: int
    payload: Dict[str, Any]
    result_summary: Dict[str, Any]
    cancel_requested: bool
    started_at: Optional[datetime]
    completed_at: Optional[datetime]
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class BatchJobListResponse(BaseModel):
    jobs: List[BatchJobResponse]
    total: int
    skip: int
    limit: int

File: frontend/src/lib/api/batch.ts
import { apiFetch } from "./client";

export type BatchOperationType = "analyze" | "score" | "tag" | "archive";
export type BatchStatus = "pending" | "running" | "completed" | "failed" | "partial" | "cancelled";

export interface BatchItemResult {
  status: "success" | "error";
  result?: unknown;
  error?: string;
}

export interface BatchJob {
  id: string;
  user_id: string;
  operation_type: BatchOperationType;
  status: BatchStatus;
  total_items: number;
  processed_items: number;
  succeeded_items: number;
  failed_items: number;
  payload: Record<string, unknown>;
  result_summary: Record<string, BatchItemResult>;
  cancel_requested: boolean;
  started_at: string | null;
  completed_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface BatchJobListResponse {
  jobs: BatchJob[];
  total: number;
  skip: number;
  limit: number;
}

export interface BatchJobCreateInput {
  operation_type: BatchOperationType;
  job_ids: string[];
  params?: Record<string, unknown>;
}

export const getBatchJobs = async (params?: {
  status?: string;
  operation_type?: string;
}): Promise<BatchJobListResponse> => {
  const search = new URLSearchParams();
  if (params?.status) search.set("status", params.status);
  if (params?.operation_type) search.set("operation_type", params.operation_type);
  const qs = search.toString();
  const response = await apiFetch(`/api/v1/batch${qs ? `?${qs}` : ""}`);
  if (!response.ok) throw new Error("Failed to fetch batch jobs");
  return response.json();
};

export const getBatchJob = async (id: string): Promise<BatchJob> => {
  const response = await apiFetch(`/api/v1/batch/${id}`);
  if (!response.ok) throw new Error("Failed to fetch batch job");
  return response.json();
};

export const createBatchJob = async (input: BatchJobCreateInput): Promise<BatchJob> => {
  const response = await apiFetch("/api/v1/batch", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(input),
  });
  if (!response.ok) {
    const err = await response.json().catch(() => ({ detail: "Failed to create batch job" }));
    throw new Error(err.detail || "Failed to create batch job");
  }
  return response.json();
};

export const retryBatchJob = async (id: string): Promise<BatchJob> => {
  const response = await apiFetch(`/api/v1/batch/${id}/retry`, { method: "POST" });
  if (!response.ok) {
    const err = await response.json().catch(() => ({ detail: "Failed to retry batch job" }));
    throw new Error(err.detail || "Failed to retry batch job");
  }
  return response.json();
};

export const cancelBatchJob = async (id: string): Promise<BatchJob> => {
  const response = await apiFetch(`/api/v1/batch/${id}/cancel`, { method: "POST" });
  if (!response.ok) {
    const err = await response.json().catch(() => ({ detail: "Failed to cancel batch job" }));
    throw new Error(err.detail || "Failed to cancel batch job");
  }
  return response.json();
};
